From: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Date: Wed, 15 Aug 2018 17:13:26 +0200
Subject: sessions: Tweak the system to make it for ubuntu

The session contains:
 - ubuntu (gnome-shell with ubuntu tweaks) in ubuntu-session package
 - unity in unity-session package
 - zarya (gnome-shell woth zarya's defaults) in zarya-session package
 - gnome-shell in the main gnome-session package.

Forwarded: Not needed
---
 data/gnome-wayland.desktop.in.in  |  2 +-
 data/gnome-xorg.desktop.in.in     |  2 +-
 data/gnome.desktop.in.in          |  2 +-
 data/meson.build                  | 43 ++++++++++++++++++++++++++++++++++++++-
 data/ubuntu-wayland.desktop.in.in |  8 ++++++++
 data/ubuntu-xorg.desktop.in.in    |  8 ++++++++
 data/ubuntu.desktop.in.in         |  8 ++++++++
 data/ubuntu.session.conf.in       |  5 +++++
 data/ubuntu.session.desktop.in.in |  4 ++++
 data/unity.desktop.in.in          |  7 +++++++
 data/unity.session.desktop.in.in  |  4 ++++
 data/wayland-sessions/meson.build |  1 +
 meson_post_install.py             |  6 +++---
 po/POTFILES.in                    |  6 ++++++
 po/POTFILES.skip                  |  6 ++++++
 15 files changed, 105 insertions(+), 7 deletions(-)
 create mode 100644 data/ubuntu-wayland.desktop.in.in
 create mode 100644 data/ubuntu-xorg.desktop.in.in
 create mode 100644 data/ubuntu.desktop.in.in
 create mode 100644 data/ubuntu.session.conf.in
 create mode 100644 data/ubuntu.session.desktop.in.in
 create mode 100644 data/unity.desktop.in.in
 create mode 100644 data/unity.session.desktop.in.in
 create mode 100644 data/wayland-sessions/meson.build

--- a/data/gnome-wayland.desktop.in.in
+++ b/data/gnome-wayland.desktop.in.in
@@ -1,7 +1,7 @@
 [Desktop Entry]
 Name=GNOME on Wayland
 Comment=This session logs you into GNOME
-Exec=@bindir@/gnome-session
+Exec=@bindir@/gnome-session --session=gnome
 TryExec=@bindir@/gnome-session
 Type=Application
 DesktopNames=GNOME
--- a/data/gnome-xorg.desktop.in.in
+++ b/data/gnome-xorg.desktop.in.in
@@ -1,7 +1,7 @@
 [Desktop Entry]
 Name=GNOME on Xorg
 Comment=This session logs you into GNOME
-Exec=@bindir@/gnome-session
+Exec=@bindir@/gnome-session --session=gnome
 TryExec=@bindir@/gnome-session
 Type=Application
 DesktopNames=GNOME
--- a/data/gnome.desktop.in.in
+++ b/data/gnome.desktop.in.in
@@ -1,7 +1,7 @@
 [Desktop Entry]
 Name=GNOME
 Comment=This session logs you into GNOME
-Exec=@bindir@/gnome-session
+Exec=@bindir@/gnome-session --session=gnome
 TryExec=@bindir@/gnome-session
 Type=Application
 DesktopNames=GNOME
--- a/data/meson.build
+++ b/data/meson.build
@@ -4,6 +4,17 @@
   desktop_plain,
   'gnome-xorg',
   'gnome-wayland',
+  'ubuntu',
+  'ubuntu-wayland',
+  'ubuntu-xorg',
+  'unity',
+  'zarya',
+  'zarya-wayland',
+  'zarya-xorg',
+]
+
+non_wayland_desktops = [
+  'unity',
 ]
 
 shell_component = {
@@ -31,6 +42,27 @@
   ],
 }
 
+shell_component += {
+  'ubuntu': shell_component[desktop_plain],
+  'zarya': shell_component[desktop_plain],
+}
+
+ubuntu_required_components = []
+ubuntu_excluded_components = [
+  'org.gnome.SettingsDaemon.UsbProtection',
+]
+
+foreach component: required_components[desktop_plain]
+  if component not in ubuntu_excluded_components
+    ubuntu_required_components += component
+  endif
+endforeach
+
+required_components += {
+  'ubuntu': ubuntu_required_components,
+  'zarya': ubuntu_required_components,
+}
+
 if enable_session_selector
   desktops += 'gnome-custom-session'
 endif
@@ -47,6 +79,7 @@
     configuration: desktop_conf
   )
 
+  all_sessions_desktop = false
   if name.endswith('-xorg')
     install_dir = session_datadir / 'xsessions'
   elif name.endswith('-wayland')
@@ -61,6 +94,7 @@
     #  join_paths(session_datadir, 'xsessions'),
     #  join_paths(session_datadir, 'wayland-sessions')
     #]
+    all_sessions_desktop = name not in non_wayland_desktops
   endif
 
   desktop_target = i18n.merge_file(
@@ -71,11 +105,24 @@
     install: true,
     install_dir: install_dir
   )
+
+  if all_sessions_desktop
+    custom_target(desktop + '-wayland',
+      command: [ 'cat', desktop_target ],
+      capture: true,
+      output: desktop + '.tmp-session',
+      install: true,
+      install_dir: session_datadir / 'wayland-sessions',
+    )
+  endif
 endforeach
 
 sessions = [
   'gnome',
-  'gnome-dummy'
+  'gnome-dummy',
+  'ubuntu',
+  'unity',
+  'zarya',
 ]
 
 foreach session: sessions
--- /dev/null
+++ b/data/ubuntu-wayland.desktop.in.in
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Name=Ubuntu on Wayland
+Comment=This session logs you into Ubuntu
+Exec=env GNOME_SHELL_SESSION_MODE=ubuntu @bindir@/gnome-session --session=ubuntu
+TryExec=@bindir@/gnome-shell
+Type=Application
+DesktopNames=ubuntu:GNOME
+X-GDM-SessionRegisters=true
--- /dev/null
+++ b/data/ubuntu-xorg.desktop.in.in
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Name=Ubuntu on Xorg
+Comment=This session logs you into Ubuntu
+Exec=env GNOME_SHELL_SESSION_MODE=ubuntu @bindir@/gnome-session --session=ubuntu
+TryExec=@bindir@/gnome-shell
+Type=Application
+DesktopNames=ubuntu:GNOME
+X-GDM-SessionRegisters=true
--- /dev/null
+++ b/data/ubuntu.desktop.in.in
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Name=Ubuntu
+Comment=This session logs you into Ubuntu
+Exec=env GNOME_SHELL_SESSION_MODE=ubuntu @bindir@/gnome-session --session=ubuntu
+TryExec=@bindir@/gnome-shell
+Type=Application
+DesktopNames=ubuntu:GNOME
+X-GDM-SessionRegisters=true
--- /dev/null
+++ b/data/ubuntu.session.conf.in
@@ -0,0 +1,5 @@
+[Unit]
+# Must be in sync with ubuntu.session
+@wants_required_components@
+
+Requires=@requires_component@.target
--- /dev/null
+++ b/data/ubuntu.session.desktop.in.in
@@ -0,0 +1,4 @@
+[GNOME Session]
+Name=Ubuntu
+# Must be in sync with gnome-session@ubuntu.target.d/ubuntu.session.conf drop-in
+RequiredComponents=@required_components@;
--- /dev/null
+++ b/data/unity.desktop.in.in
@@ -0,0 +1,7 @@
+[Desktop Entry]
+Name=Unity
+Comment=This session logs you into Unity
+Exec=/usr/libexec/run-systemd-session unity-session.target
+TryExec=@bindir@/unity
+Type=Application
+DesktopNames=Unity:Unity7:ubuntu
--- /dev/null
+++ b/data/unity.session.desktop.in.in
@@ -0,0 +1,4 @@
+[GNOME Session]
+Name=Unity
+RequiredComponents=unity-settings-daemon;
+DesktopName=Unity:Unity7:ubuntu
--- /dev/null
+++ b/data/wayland-sessions/meson.build
@@ -0,0 +1 @@
+
--- a/meson_post_install.py
+++ b/meson_post_install.py
@@ -1,5 +1,6 @@
 #!/usr/bin/env python3
 
+import glob
 import os
 import shutil
 import subprocess
@@ -21,6 +22,5 @@
 if not os.path.exists(dst_dir):
   os.makedirs(dst_dir)
 
-src = os.path.join(install_root, 'xsessions', 'gnome.desktop')
-dst = os.path.join(dst_dir, 'gnome.desktop')
-shutil.copyfile(src, dst)
+for i in glob.glob(os.path.join(dst_dir, '*.tmp-session')):
+  os.rename(i, i[:-len('.tmp-session')])
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -8,6 +8,16 @@
 data/gnome-xorg.desktop.in.in
 data/org.gnome.SessionManager.gschema.xml.in
 data/session-selector.ui
+data/ubuntu.desktop.in.in
+data/ubuntu-wayland.desktop.in.in
+data/ubuntu-xorg.desktop.in.in
+data/ubuntu.session.desktop.in.in
+data/unity.desktop.in.in
+data/unity.session.desktop.in.in
+data/zarya.desktop.in.in
+data/zarya-wayland.desktop.in.in
+data/zarya-xorg.desktop.in.in
+data/zarya.session.desktop.in.in
 gnome-session/gsm-fail-whale-dialog.c
 gnome-session/gsm-manager.c
 gnome-session/gsm-process-helper.c
--- a/po/POTFILES.skip
+++ b/po/POTFILES.skip
@@ -7,3 +7,13 @@
 data/gnome.session.desktop.in
 data/gnome-wayland.desktop.in
 data/gnome-xorg.desktop.in
+data/ubuntu.desktop.in
+data/ubuntu-wayland.desktop.in
+data/ubuntu-xorg.desktop.in
+data/ubuntu.session.desktop.in
+data/unity.desktop.in
+data/unity.session.desktop.in
+data/zarya.desktop.in.in
+data/zarya-wayland.desktop.in.in
+data/zarya-xorg.desktop.in.in
+data/zarya.session.desktop.in.in
--- /dev/null
+++ b/data/zarya-wayland.desktop.in.in
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Name=zarya on Wayland
+Comment=This session logs you into zarya
+Exec=env GNOME_SHELL_SESSION_MODE=zarya @bindir@/gnome-session --session=zarya
+TryExec=@bindir@/gnome-shell
+Type=Application
+DesktopNames=zarya:GNOME
+X-GDM-SessionRegisters=true
--- /dev/null
+++ b/data/zarya-xorg.desktop.in.in
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Name=zarya on Xorg
+Comment=This session logs you into zarya
+Exec=env GNOME_SHELL_SESSION_MODE=zarya @bindir@/gnome-session --session=zarya
+TryExec=@bindir@/gnome-shell
+Type=Application
+DesktopNames=zarya:GNOME
+X-GDM-SessionRegisters=true
--- /dev/null
+++ b/data/zarya.desktop.in.in
@@ -0,0 +1,8 @@
+[Desktop Entry]
+Name=zarya
+Comment=This session logs you into zarya
+Exec=env GNOME_SHELL_SESSION_MODE=zarya @bindir@/gnome-session --session=zarya
+TryExec=@bindir@/gnome-shell
+Type=Application
+DesktopNames=zarya:GNOME
+X-GDM-SessionRegisters=true
--- /dev/null
+++ b/data/zarya.session.conf.in
@@ -0,0 +1,5 @@
+[Unit]
+# Must be in sync with zarya.session
+@wants_required_components@
+
+Requires=@requires_component@.target
--- /dev/null
+++ b/data/zarya.session.desktop.in.in
@@ -0,0 +1,4 @@
+[GNOME Session]
+Name=zarya
+# Must be in sync with gnome-session@zarya.target.d/zarya.session.conf drop-in
+RequiredComponents=@required_components@;
